crates/rollup/src/driver/mod.rs

```diff
@@ -118,12 +155,80 @@ where
     }
 
     /// Advance the pipeline to the next L2 block.
-    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> Result<()> {
-        let (l2_tip, l1_origin) = self.cursor.tip();
-        let _ = pipeline.step(l2_tip).await;
-        self.cursor.advance(l1_origin, l2_tip);
+    ///
+    /// Returns `true` if the pipeline can move forward again, `false` otherwise.
+    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> bool {
+        let l2_tip = self.cursor.tip();
+
+        match pipeline.step(l2_tip).await {
+            StepResult::PreparedAttributes => trace!("Perpared new attributes"),
+            StepResult::AdvancedOrigin => trace!("Advanced origin"),
+            StepResult::OriginAdvanceErr(err) => warn!("Could not advance origin: {:?}", err),
+            StepResult::StepFailed(err) => match err {
+                StageError::NotEnoughData => debug!("Not enough data to advance pipeline"),
+                _ => error!("Error stepping derivation pipeline: {:?}", err),
+            },
+        }
+
+        let derived_attributes = if let Some(attributes) = pipeline.peek() {
+            match self.validator.validate(attributes).await {
+                Ok(true) => {
+                    trace!("Validated payload attributes");
+                    pipeline.next().expect("Peeked attributes must be available")
+                }
+                Ok(false) => {
+                    error!("Failed payload attributes validation");
+                    // TODO: should we advance the pipeline iterator here?
+                    return false;
+                }
+                Err(err) => {
+                    error!("Error while validating payload attributes: {:?}", err);
+                    return false;
+                }
+            }
+        } else {
+            debug!("No attributes available to validate");
+            return false;
+        };
 
-        unimplemented!()
+        let derived = derived_attributes.parent.block_info.number + 1;
+        // Q: can we use the origin from `pipeline.origin()` instead of fetching it from L1?
```

Comment:
The l1 origin is the block # and hash - we can advance the number but we won't know the new hash. Since these are all cached though, isn't this a super lightweight op? It shouldn't add any latency really as I understand

Reply:
ok makes sense!

> Since these are all cached though, isn't this a super lightweight op?

They are cached depending on the L1 chain provider: right now the StandaloneDriver is using an alloy provider.

The Standalone driver DOES have L1 blocks cached, but they are in the Context itself, and not in the chain provider. Right now these are separate components. I realize it's confusing so here is a schematic:

- ExEx mode:
  - L1 notifications come from the ExEx, and L1 blocks get cached in the in-memory provider
  - All calls to the provider will return the cached data
- Standalone mode:
  - L1 notifications come from the StandaloneContext which is just a L1 block event subscriber
  - The L1 provider is online with no cache (currently)
  - So when fetching from the L1 provider, we call the RPC even if the StandaloneContext might have that block cached

This can be improved in a number of ways for which a new issue is probably best!